.github/workflows/test.yml

```diff
@@ -0,0 +1,57 @@
+---
+## Test with multiple ruby images and frameworks
+name: test
+
+on:
+  push:
+    branches:
+      - main
+    tags:
+      - v*.*.*
+  pull_request:
+    branches:
+      - main
+
+permissions:
+  contents: read
+
+jobs:
+  test:
+    strategy:
+      fail-fast: false
+      matrix:
+        ruby-image:
+          - 'ruby:2.7'
+          - 'ruby:2.6'
+          - 'ruby:2.5'
+          - 'ruby:2.4'
+          - 'ruby:2.3'
+          - 'jruby:9.2'
+        framework:
+          - 'rails-6.1'
+          - 'rails-6.0'
+          - 'rails-5.2'
+          - 'sinatra-2.0'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: '2.7.2'
+      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@current
+        if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository # Only docker-login if the PR is not from a fork.
```

Comment:
FYI, I added this condition because secrets.* are not available if the PR is from a fork.

Reply:
Or should we just omit the whole docker-login step. WDYT?

Reply:
AFAIS, this action shouldn't be required since we build the docker image before running tests.

Reply:
The images are built during the `bin/dev` command and during the build the base image is still pulled.

<details>
  <summary>See GH Actions log</summary>
  
  ```
  2.7: Pulling from library/ruby
  f2f58072e9ed: Already exists
  5c8cfbf51e6e: Already exists
  aa3a609d[15](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:16)79: Already exists
  094e7d9bb04e: Already exists
  2cbfd734f382: Already exists
  29360[16](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:17)f2004: Pulling fs layer
  b99019e2c9ef: Pulling fs layer
  58c283[17](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:18)6d31: Pulling fs layer
  58c283176d31: Verifying Checksum
  58c283176d31: Download complete
  2936016f2004: Verifying Checksum
  2936016f2004: Download complete
  b990[19](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:20)e2c9ef: Verifying Checksum
  b99019e2c9ef: Download complete
  2936016f[20](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:21)04: Pull complete
  b99019e2c9ef: Pull complete
  58c[28](https://github.com/elastic/ecs-logging-ruby/actions/runs/3749897639/jobs/6368925770#step:5:29)3176d31: Pull complete
  ```
</details>

Hence, on every workflow run, it might pull the image 24 times. 

Do you know by chance how likely it is, that we will reach the anonymous docker rate limit?

Reply:
Or can this also affect other workflows in other repos?

Reply:
I think that worth to explore how to cache the Docker images. If we cache the docker images in GitHub Actions we do not have to worry about the Docker Hub limit. It is something we will hit on more repositories.  

https://github.com/orgs/community/discussions/26330#discussioncomment-3251498
https://github.com/marketplace/actions/docker-layer-caching

Reply:
use services to grab the docker images is another option https://docs.github.com/en/actions/using-containerized-services/about-service-containers

Reply:
https://github.com/elastic/ecs-logging-ruby/pull/28/files/79eb499b917161f2c43250244f136d77493f81f3..8e2faf159773348b987ff7cf9dd667916d03b414

The docker-layer caching seems to work fine.

```
Step 1/20 : ARG RUBY_IMAGE
Step 2/20 : FROM ${RUBY_IMAGE}
 ---> f172588bbb0b
Step 3/20 : ARG USER_ID_GROUP
 ---> Using cache
...
 ```
Source: https://github.com/elastic/ecs-logging-ruby/actions/runs/3757138181/jobs/6384047047

In this [workflow run](https://github.com/elastic/ecs-logging-ruby/actions/runs/3757138181/attempts/1) you can see it saving into cache in the `Post Run`

Reply:
> The images are built during the `bin/dev` command and during the building the base image is still pulled.
> 
> See GH Actions log
> Hence, on every workflow run, it might pull the image 24 times.
> 
> Do you know by chance how likely it is, that we will reach the anonymous docker rate limit in GH actions?
> 
> However, looking at the highest PR number there doesn't seem to be much traffic on this repo. Hence, I assume we will survive without the docker login.
> 
> WDYT?

I missed the target registry URL. As you mentioned, caching docker layers should be enough to avoid rate limit.